js/src/common/utils/ItemList.ts

```diff
@@ -1,35 +1,51 @@
-class Item {
-  content: any;
+class Item<T> {
+  content:
+    | T
+    | (T & {
+        /**
+         * Set when calling `.toArray()`
+         */
+        itemName: string;
```

Comment:
I don't think we should mutate internal state when generating toArray output. Could we avoid doing so, and as a result, simplify these types?